### ISSUE TITLE
Keras speed slow on VGG-16

### ISSUE THREAD

**Issue by aizvorski** [fchollet/keras#3511]
I was comparing Keras with Caffe and found that Keras is 3.5x slower using Theano or 5x slower using TensorFlow.  This is using CUDA 8, cuDNN 5.0 on GTX 1080.  

I packaged up the comparison here: https://github.com/aizvorski/caffe-vs-keras

As far as I can tell all three implementations are working correctly (they converge if run outside of a benchmark), and have no problems using 100% of the GPU as shown by nvidia-smi dmon.  

Also as far as I can tell there is nothing unusual about my implementation of VGG-16 (the Caffe version is from the model zoo, the Keras version is from https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3).

This is surprising because I would expect all implementations running on top of cuDNN to have approximately the same speed, barring something quite unusual.

Is this expected? Could it be a performance bug?

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by fchollet** [comment]
Are you aware of: https://github.com/fchollet/deep-learning-models

Your network has a ton of necessary operations (padding). That's why it's slow.

**Comment by fchollet** [comment]
Oh yeah, and in the TF case you are not using the native TF dim ordering. So it would be even slower.

**Comment by fchollet** [comment]
TL;DR: use the models I provide, they are much better optimized.

**Comment by aizvorski** [comment]
@fchollet  Thank you, these are good pointers.  I used the models you pointed me to, and fixed the TF dim ordering issue.  My repo has been updated to reflect this.  This helped but did not fix the problem.

Now Keras+TF is 1.9x slower, and Keras+Theano is 2.7x slower than Caffe. 

Is this expected?  If not, perhaps you can reopen this issue.

**Comment by aizvorski** [comment]
@fchollet The Caffe benchmark was doing less work since it wasn't updating weights, just calculating gradients.  Fixed.  

Now Keras+TF is 1.2x slower, and Keras+Theano is 1.7x slower.  The TF speed is not too far from parity.

I have confirmed that Lasagne+Theano is almost exactly the same speed as Keras+Theano, so this seems to be more of a Theano issue than a Keras issue.

**Comment by fchollet** [comment]
You should try comparing the generated TF graphs for the pure-TF version
and the TF version. IMO they should be the exact same (or close enough) and
thus there should be no speed difference.

You can visualize them in TensorBoard, for instance.
